src/shm_main.c

```diff
@@ -768,6 +813,273 @@ sr_shmmain_createunlock(int shm_lock)
     }
 }
 
+/**
+ * @brief Determine whether the connection specified by cid is alive.
+ *
+ * @param[in]  cid Connection ID to check
+ * @param[out] conn_alive Set to non-zero if the connection is alive, 0 otherwise.
+ * @return err_info, NULL on success.
+ */
+sr_error_info_t *
+sr_shmmain_check_conn_lock(sr_cid_t cid, int *conn_alive) {
+    sr_error_info_t *err_info = NULL;
+    struct flock fl;
+    int fd = -1;
+    char *path = NULL;
+    int rc = 0;
+    conn_list_entry *ptr = NULL;
+
+    assert(conn_alive);
+
+    (*conn_alive) = 0;
+
+    /* Connection ID NONE is reserved and never used. */
+    if (SR_CONN_ID_NONE == cid) {
+        return NULL;
+    }
+
+    /* If the connection is owned by this process a check using flock which
+     * requires an open/close would release the lock. Check if the CID is a
+     * connection owned by this process and return status before we do an
+     * open().
+     */
+    if (NULL != conn_head) {
+        if ( (err_info = sr_mlock(conn_list_lock, 1000, __func__))) {
+            return err_info;
+        }
+        for (ptr = conn_head; ptr != NULL; ptr = ptr->_next) {
+            if (cid == ptr->cid) {
+                *conn_alive = 1;
+                sr_munlock(conn_list_lock);
+                return NULL;
+            }
+        }
+        sr_munlock(conn_list_lock);
+    }
+
+    sr_shmmain_get_conn_lockfile_path(cid, &path);
+
+    /* Open the file to test the lock */
+    fd = open(path, O_RDWR);
+    free(path);
+    if (-1 == fd) {
+        /* The file does not exist in which case there is no connection established */
+        if (ENOENT == errno) {
+            (*conn_alive) = 0;
+            return NULL;
+        }
+        SR_ERRINFO_SYSERRNO(&err_info, "open");
+        return err_info;
+    }
+
+    /* Check the lock */
+    memset(&fl, 0, sizeof fl);
+    fl.l_whence = SEEK_SET;
+    fl.l_start = 0;
+    fl.l_len = 0; /* length of 0 is entire file */
+    fl.l_type = F_WRLCK;
+    rc = fcntl(fd, F_GETLK, &fl);
+    close(fd); /* this will release the lock if we hold it, thus the list above. */
```

Comment:
Connected to the comment on line 56, is there more to the list? If this unlocks the file lock, is the list used for relocking it? I do not see anything like that so I am wondering where it gets locked again.

Reply:
Based on the scanning of the list above (in this function) we will never close an FD which is used in tracking a connection owned by our process.  That is done only in the deallocate function.

Reply:
Right.

Reply:
Done